core/src/main/scala/slamdata/AutoBump.scala

```diff
@@ -0,0 +1,369 @@
+          "GIT_COMMITTER_NAME" -> s"Precog Bot ($authorRepository)",
+          "GIT_COMMITTER_EMAIL" -> "bot@precog.com").void.attempt
+      } yield result.leftMap(_ => NoChangesError)
+    }
+
+    def tryPush(dirFile: File, branchName: String): IO[Either[Warnings, Unit]] = {
+      run(s"git push origin $branchName", log, merge = true, workingDir = Some(dirFile))
+        .void
+        .attempt
+        .map(_.leftMap(_ => PushError))
+    }
+
+    def ifOldest(pullRequest: PullRequestDraft): IO[Either[Warnings, PullRequestDraft]] = for {
+      _ <- markReady(pullRequest)
+      _ <- IO(log.info(s"Marked $owner/$repoSlug#${pullRequest.number} ready for review"))
+    } yield pullRequest.asRight[Warnings]
+
+    def ifNotOldest(oldest: PullRequestDraft, pullRequest: PullRequestDraft): IO[Either[Warnings, PullRequestDraft]] = for {
+      _ <- close(pullRequest, s"${pullRequest.title} (preceded by #${oldest.number})")
+      _ <- IO(log.info(s"Closed $owner/$repoSlug#${pullRequest.number}, preceded by $owner/$repoSlug#${oldest.number}"))
+      _ <- deleteBranch(pullRequest)
+      _ <- IO(log.info(s"Removed branch ${pullRequest.base.map(_.ref)} from $owner/$repoSlug"))
+    } yield NotOldest(oldest, pullRequest).asLeft[PullRequestDraft]
+
+    // TODO: Update pull request description?
+    def createOrUpdatePullRequest(branchName: String,
+                                  changes: List[String],
+                                  changeLabel: ChangeLabel,
+                                  maybePullRequest: Option[PullRequestDraft]): IO[Either[Warnings, PullRequestDraft]] = {
+      for {
+        pullRequest <- maybePullRequest.fold {
+          draftPullRequest(authorRepository, branchName, changes.mkString("\n"))
+            .flatTap(pullRequest => IO(log.info(s"Opened $owner/$repoSlug#${pullRequest.number}")))
+        }(IO.pure)
+        labels <- getLabels(pullRequest.number)
+        prChangeLabels = labels.flatMap(label => ChangeLabel(label.name)).toSet
+        highestChange = (prChangeLabels + changeLabel).max
+        lowerChanges = (prChangeLabels - highestChange).toList
+        _ <- if (prChangeLabels.contains(highestChange)) IO.unit else assignLabel(highestChange.label, pullRequest)
+        _ <- lowerChanges.traverse(change => removeLabel(pullRequest, change.label))
+        oldestPullRequest <- getOldestAutoBumpPullRequest
+        res <- oldestPullRequest
+          .filter(_.number == pullRequest.number)
+          .fold(ifOldest(pullRequest))(ifNotOldest(_, pullRequest))
+      } yield res
+    }
+
+    val app = for {
+      (dir, branchName, maybePullRequest, changes, label) <- EitherT(tryUpdateDependencies)
+      _ <- EitherT(tryCommit(dir))
+      _ <- EitherT(tryPush(dir, branchName))
+      pullRequest <- EitherT(createOrUpdatePullRequest(branchName, changes, label, maybePullRequest))
+    } yield pullRequest
+
+    app.leftSemiflatMap(_.warn(log)).value.unsafeRunSync().isRight
```

Comment:
I'm considering returning `IO` here, to make it testable. Thoughts?

Reply:
In general, you want to push `IO` as far out to the leaves as possible. In this case, that would mean returning `IO` (i.e. *not* using `unsafe` anything) all the way to the setting/task definitions, where you would of course need to use the `unsafe` function to run it. There isn't *that* much of a difference in terms of testability, but it does help a lot in terms of reasoning when and where effects are run.

So tldr, I do think returning `IO` here would be better.